ddl/sanity_check.go

```diff
@@ -0,0 +1,243 @@
+			panic("should not happened")
+		}
+		checkRangeCntByTableIDs(physicalTableIDs, cnt)
+	case model.ActionDropTablePartition, model.ActionTruncateTablePartition:
+		var physicalTableIDs []int64
+		if err := job.DecodeArgs(&physicalTableIDs); err != nil {
+			panic("should not happened")
+		}
+		if len(physicalTableIDs) != int(cnt) {
+			panic("should not happened")
+		}
+	case model.ActionAddIndex, model.ActionAddPrimaryKey:
+		var indexID int64
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&indexID, &partitionIDs); err != nil {
+			panic("should not happened")
+		}
+		checkRangeCntByTableIDs(partitionIDs, cnt)
+	case model.ActionDropIndex, model.ActionDropPrimaryKey:
+		var indexName interface{}
+		var indexID int64
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&indexName, &indexID, &partitionIDs); err != nil {
+			panic("should not happened")
+		}
+		checkRangeCntByTableIDsAndIndexIds(partitionIDs, []int64{indexID}, cnt)
+	case model.ActionDropIndexes:
+		var indexIDs []int64
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&[]model.CIStr{}, &[]bool{}, &indexIDs, &partitionIDs); err != nil {
+			panic("should not happened")
+		}
+		checkRangeCntByTableIDsAndIndexIds(partitionIDs, indexIDs, cnt)
+	case model.ActionDropColumn:
+		var colName model.CIStr
+		var indexIDs []int64
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&colName, &indexIDs, &partitionIDs); err != nil {
+			panic("should not happened")
+		}
+		checkRangeCntByTableIDsAndIndexIds(partitionIDs, indexIDs, cnt)
+	case model.ActionDropColumns:
+		var colNames []model.CIStr
+		var ifExists []bool
+		var indexIDs []int64
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&colNames, &ifExists, &indexIDs, &partitionIDs); err != nil {
+			panic("should not happened")
+		}
+		checkRangeCntByTableIDsAndIndexIds(partitionIDs, indexIDs, cnt)
+	case model.ActionModifyColumn:
+		var indexIDs []int64
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&indexIDs, &partitionIDs); err != nil {
+			panic("should not happened")
+		}
+		checkRangeCntByTableIDsAndIndexIds(partitionIDs, indexIDs, cnt)
+	}
+}
+
+// checkHistoryJobInTest does some sanity check to make sure something is correct after DDL complete.
+// It's only check during the test environment, so it would panic directly.
+// These checks may be controlled by configuration in the future.
+func (d *ddl) checkHistoryJobInTest(ctx sessionctx.Context, historyJob *model.Job) {
+	if !(flag.Lookup("test.v") != nil || flag.Lookup("check.v") != nil) {
+		return
+	}
+
+	// Check delete range.
+	// Keep it the same as finishDDLJob.
+	if !historyJob.IsCancelled() {
+		switch historyJob.Type {
+		case model.ActionAddIndex, model.ActionAddPrimaryKey:
+			if historyJob.State != model.JobStateRollbackDone {
+				break
+			}
+			fallthrough
+		case model.ActionDropSchema, model.ActionDropTable, model.ActionTruncateTable, model.ActionDropIndex, model.ActionDropPrimaryKey,
```

Comment:
Do we need to add a function to identify tasks of this type？Because there are a few places to use, in case of missing.

Reply:
I don't have a good idea, any suggestion?

Reply:
Do we add a function like `IsRollbackDone`?